services/src/main/java/org/keycloak/services/client/DefaultClientService.java

```diff
@@ -36,14 +36,28 @@ public Stream<ClientRepresentation> getClients(RealmModel realm, ClientProjectio
     }
 
     @Override
-    public ClientRepresentation createClient(RealmModel realm, ClientRepresentation client) throws ServiceException {
-        if (realm.getClientByClientId(client.getClientId()) != null) {
-            throw new ServiceException("Client already exists", Response.Status.CONFLICT);
+    public CreateOrUpdateResult createOrUpdate(RealmModel realm, ClientRepresentation client, boolean allowUpdate)
+            throws ServiceException {
+        boolean created = false;
+        ClientModel model = realm.getClientByClientId(client.getClientId());
+        if (model != null) {
+            if (!allowUpdate) {
+                throw new ServiceException("Client already exists", Response.Status.CONFLICT);
+            }
+            // TODO: make sure that removal is the best way to do this
```

Comment:
It's probably not. Just from the top of my head, we might accidentally delete some entities with foreign keys. But there might be more caveats.

Reply:
As long as the overlay of the model state is complete, we won't need to use a delete - based upon the current mapper prototype there will need to be some changes to support that.

Reply:
Right. But we can tackle that as a follow-up.